crates/shim-sgx/src/handler/mod.rs

```diff
@@ -179,15 +181,60 @@ impl guest::Handler for Handler<'_> {
         Ok(())
     }
 
-    // Until EDMM, we can't change any page permissions.
-    // What you get is what you get. Fake success.
     fn mprotect(
         &mut self,
         _platform: &impl Platform,
-        _addr: NonNull<c_void>,
-        _len: c_size_t,
-        _prot: c_int,
+        addr: NonNull<c_void>,
+        length: c_size_t,
+        prot: c_int,
     ) -> sallyport::Result<()> {
+        let addr = addr.as_ptr() as usize;
+        let length = (length + 0xfff) & !0xfff;
+
+        if addr & 0xfff != 0 || length == 0 {
+            return Err(EINVAL);
+        }
+
+        // Given that the hardware asserts on !W^R, the function, by practical
+        // means, only RWX is disregarded from non-empty subsets of permissions.
+        if prot != PROT_READ && prot != (PROT_READ | PROT_WRITE) && prot != (PROT_READ | PROT_EXEC)
+        {
+            debugln!(self, "({:>#016x}, {:>#016x}, {:>#02x})", addr, length, prot);
+            return Err(EACCES);
+        }
+
+        let mut heap = HEAP.write();
+
+        if heap.contains(Region::new(Address::new(addr), Address::new(addr + length))) == None {
+            return Err(ENOMEM);
+        }
+
+        // TODO: Change the backend to zero the permissions instead of
+        // restricting them to 'read' (SGX2 series v4 issue).
+        self.reset_sgx_permissions(NonNull::new(addr as *mut _).unwrap(), length)?;
+
+        for i in 0..length / Page::SIZE {
+            let virt_addr = VirtAddr::new((addr + i * Page::SIZE) as u64);
+            let page_addr = PageAddr::from_start_address(virt_addr).unwrap();
+
+            // TODO: Remove `Flags::READ | ` when possible (SGX2 series v4
+            // issue)
+            Class::Regular
+                .info(Flags::READ | Flags::RESTRICTED)
+                .accept(page_addr)
+                .unwrap();
+
+            Class::Regular
+                .info(Flags::from_bits_truncate(prot as u8))
+                .extend_permissions(page_addr);
```

Comment:
`.extend_permissions()` seems like a long function name. Should this be renamed to `.extend()`?

Reply:
I agree with this but can it be done after this PR as it needs an update to `sgx` crate?

Reply:
I can create a debt issue for `sgx` crate for this, should  I do that?

Reply:
Renamed but now the PR depends on my local tree, i.e. this needs to be merged: https://github.com/enarx/sgx/pull/91